src/a2rl/_dataframe.py

```diff
@@ -679,6 +679,110 @@ def add_value(
+        episode_identifier: identifier in the 'self' dataframe that allows for groupby and
+            identifies a process.
+
+    Returns:
+        This dataframe, modified with an additional ``value_col`` column. This return value is
+        provided to facilitate chaining as-per the functional programming style.
+    """
+    self._check_add_value_args(value_col, override)
+    if len(self._rewards) == 1:
+        df = self
+    else:
+        df = WiDataFrame(
+            self,
+            states=self.states,
+            actions=self.actions,
+            rewards=self.rewards[:1],
+        )
+    df_t = wi.DiscreteTokenizer(n_bins=50).fit_transform(df.trim())
+
+    def calc_q_values(sub, df, x, alpha, gamma, sarsa=sarsa):
+        # print(x.name)
+        if sub.name == 0:
+            return pd.Series(None, index=sub.index)
+
+        df = df.loc[sub.index]
+        x = x.loc[sub.index]
+
+        # Temp df with only three columns: _state, _action, _reward
+        df_t = pd.concat(  # type: ignore[assignment]
+            [
+                x[x.states].astype(str).apply("_".join, axis=1).astype("category").cat.codes,
+                x[x.actions].astype(str).apply("_".join, axis=1).astype("category").cat.codes,
+                df[df.rewards],  # .reset_index(drop=False),
+            ],
+            axis=1,
+            copy=False,
+        )
+
+        df_t.columns = ["_state", "_action", "_reward"]  # type: ignore[assignment]
+
+        q_table = np.zeros([df_t["_state"].nunique(), df_t["_action"].nunique()])
+
+        iterations = 10
+
+        for n in range(iterations):
+            for i in range(df_t.index[0], df_t.index[-1]):
+                state = int(df_t.loc[i, "_state"])
+                next_state = int(df_t.loc[i + 1, "_state"])
+                action = int(df_t.loc[i, "_action"])
+                reward = df_t.loc[i + 1, "_reward"]
+                old_value = q_table[state, action]
+
+                if sarsa:
+                    next_value = q_table[next_state, np.argmax(q_table[next_state])]
+                    new_value = (1 - alpha) * old_value + alpha * (
+                        reward + gamma * next_value - old_value
+                    )
+                else:
+                    next_max = np.max(q_table[next_state])
+                    new_value = (1 - alpha) * old_value + alpha * (
+                        reward + gamma * next_max - old_value
+                    )
+
+                q_table[state, action] = new_value
+
+        return pd.Series(
+            q_table[df_t["_state"].astype(int), df_t["_action"].astype(int)],
+            index=df.index,
+        )
+
+    x = df.groupby(episode_identifier).apply(lambda x: calc_q_values(x, df, df_t, alpha, gamma))
+    x2 = x.reset_index(level=0)
+    x2.rename({0: "value"}, axis=1, inplace=True)
+    x2.drop([episode_identifier], axis=1, inplace=True)
+    merged = pd.merge(df, x2, how="left", left_index=True, right_index=True)
+    return merged
```

Comment:
Propose to change the behavior to modify in-place -- the same semantic in the single-episode counterpart.

```suggestion
    # len_before_modif = self.shape[0]
    self[value_col] = merged[value_col]  # Make sure index matches
    # assert self.shape[0] == len_before_modif
    return self
```

However, if this breaks too much, I'm happy to accept this version first once the docstring (description & returns) is updated to say that a new dataframe is returned.